src/System.Private.CoreLib/shared/System/Marvin.cs

```diff
@@ -28,73 +28,194 @@ internal static partial class Marvin
             {
-                case 4:
-                    p0 += Unsafe.ReadUnaligned<uint>(ref Unsafe.AddByteOffset(ref data, byteOffset));
-                    Block(ref p0, ref p1);
-                    goto case 0;
-
-                case 0:
-                    p0 += 0x80u;
-                    break;
-
-                case 5:
-                    p0 += Unsafe.ReadUnaligned<uint>(ref Unsafe.AddByteOffset(ref data, byteOffset));
-                    byteOffset += 4;
-                    Block(ref p0, ref p1);
-                    goto case 1;
-
-                case 1:
-                    p0 += 0x8000u | Unsafe.AddByteOffset(ref data, byteOffset);
-                    break;
-
-                case 6:
-                    p0 += Unsafe.ReadUnaligned<uint>(ref Unsafe.AddByteOffset(ref data, byteOffset));
-                    byteOffset += 4;
-                    Block(ref p0, ref p1);
-                    goto case 2;
-
-                case 2:
-                    p0 += 0x800000u | Unsafe.ReadUnaligned<ushort>(ref Unsafe.AddByteOffset(ref data, byteOffset));
-                    break;
-
-                case 7:
-                    p0 += Unsafe.ReadUnaligned<uint>(ref Unsafe.AddByteOffset(ref data, byteOffset));
-                    byteOffset += 4;
-                    Block(ref p0, ref p1);
-                    goto case 3;
-
-                case 3:
-                    p0 += 0x80000000u | (((uint)(Unsafe.AddByteOffset(ref data, byteOffset + 2))) << 16) | (uint)(Unsafe.ReadUnaligned<ushort>(ref Unsafe.AddByteOffset(ref data, byteOffset)));
-                    break;
-
-                default:
-                    Debug.Fail("Should not get here.");
-                    break;
+                partialResult >>= 8; // make some room for the 0x80 byte
+                partialResult |= 0x80000000u; // put the 0x80 byte at the beginning
+                partialResult >>= (~count << 3) & 0x1F; // shift out all previously consumed bytes
             }
+            else
+            {
+                partialResult <<= 8; // make some room for the 0x80 byte
+                partialResult |= 0x80u; // put the 0x80 byte at the end
+                partialResult <<= (~count << 3) & 0x1F; // shift out all previously consumed bytes
+            }
+
+        DoFinalRoundsAndReturn:
 
+            // Now that we've computed the final partial result, merge it in and run two rounds of
+            // the block function to finish out the Marvin algorithm.
+
+            p0 += partialResult;
             Block(ref p0, ref p1);
             Block(ref p0, ref p1);
 
             return (int)(p1 ^ p0);
+
+        InputTooSmallToEnterMainLoop:
+
+            // We can't run the main loop, but we might still have 4 or more bytes available to us.
+            // If so, jump to the 4 .. 7 bytes logic immediately after the main loop.
+
+            if ((count & 4) != 0)
+            {
+                goto Between4And7BytesRemain;
+            }
```

Comment:
Why have this check here rather than as part of the original check at the entrance to the method, e.g.
```C#
if ((uint)count < 8)
{
    if ((count & 4) != 0) goto Between4And7BytesRemain;
    goto InputTooSmallToEnterMainLoop;
}
```
?

Reply:
I had intended to reduce the total number of instructions that ran before entering the main loop, but after making this change and rerunning the test I see no measurable difference. If the consensus is that putting the check here is cleaner I can definitely make the change.